cluster/cluster.go

```diff
@@ -155,3 +157,50 @@ LOOP:
 
 	return answer, nil
 }
+
+// MembersForSpeculativeQuery returns a prioritized list of nodes for each shard group
+// TODO - this assumes that the partition set for each node is perfectly aligned
+func MembersForSpeculativeQuery() (map[int32][]Node, error) {
+	thisNode := Manager.ThisNode()
+	allNodes := Manager.MemberList()
+	membersMap := make(map[int32][]Node)
+
+	// If we are running in single mode, just return thisNode
+	if Mode == ModeSingle {
+		membersMap[0] = []Node{thisNode}
+		return membersMap, nil
+	}
+
+	peerPartitions := 0
+
+	// store the available nodes for each partition group
+	for _, member := range allNodes {
+		if !member.IsReady() {
+			continue
+		}
+		memberStartPartition := member.GetPartitions()[0]
+
+		if _, ok := membersMap[memberStartPartition]; !ok {
```

Comment:
This seems to determine whether a member is already in the map or not based on its first partition. But I can't see where we are sorting the partitions, if two MTs of a pair are configured to have the same partitions but they are specified in a different order wouldn't this break?

Reply:
That's true. IMO, it seems like the partitions should be sorted at start up. I could add a sort to the `SetPartitions` function.

Reply:
I agree that sorting at startup makes the most sense, that way it only needs to be done once

Reply:
I just noticed that in cases where the partition ids are not sorted in the config it is important to first update to this version of MT without activating speculative queries, and only once all MTs are on this version enable speculative queries. Otherwise querying might temporarily be broken until all MTs are updated because the older ones still returned their partition IDs unsorted. But i guess there is nothing we can do about that.

Reply:
Well, we could sort the partitions of the cluster peers when we get them. As it stands right now, with or without speculation enabled, the partitions will need to be sorted. That is, unless I change the code as I mentioned in https://github.com/grafana/metrictank/pull/956#issuecomment-406676801